src/main/java/com/endava/internship/collections/StudentList.java

```diff
@@ -1,145 +1,392 @@
+            }
+        return (T[]) newResizedArray;
+        }
     }
 
     @Override
-    public boolean add(Student student) {
-        //TODO
-        return false;
+    public boolean add(T t) {
+        if (size == elements.length) {
+            resizeArray();
+        }
+        elements[size++] = t;
+        return true;
     }
 
     @Override
-    public boolean remove(Object o) {
-        //TODO
-        return false;
+    public boolean remove(Object t) {
+        if (!this.contains(t)) {
+            return false;
+        }
+        Object[] elements = decreasedArray(size);
+        int indexOfDeletedObject = this.indexOf(t);
+        copyArray(indexOfDeletedObject, elements);
+        return true;
     }
 
     @Override
     public void clear() {
-        //TODO
+        for (int i = 0; i < size; i++)
+            elements[i] = null;
+        size = 0;
     }
 
     @Override
-    public Student get(int i) {
-        //TODO
-        return null;
+    public T get(int i) {
+        if (i >= 0 && i <= size) {
+            return (T) elements[i];
+        } else {
+            throw new IndexOutOfBoundsException();
+        }
     }
 
     @Override
-    public Student set(int i, Student student) {
-        //TODO
-        return null;
+    public T set(int i, T t) {
+        if (i >= 0 && i < size) {
+            return (T) (elements[i] = t);
+        } else {
+            throw new IndexOutOfBoundsException();
+        }
     }
 
     @Override
-    public void add(int i, Student student) {
-        //TODO
+    public void add(int i, T t) {
+        if (i < 0 && i > size) {
+            throw new IndexOutOfBoundsException();
+        }
+        if (size == elements.length) {
+            resizeArray();
+        }
+        for (int start = size; start > i; start--) {
+            elements[start] = elements[start - 1];
+        }
+        elements[i] = t;
     }
 
     @Override
-    public Student remove(int i) {
-        //TODO
-        return null;
+    public T remove(int i) {
+        if (i < 0 || i >= size) {
+            throw new IndexOutOfBoundsException();
+        }
+        final Object elementToDelete = elements[i];
+        Object[] elements = decreasedArray(size);
+        copyArray(i, elements);
+        return (T) elementToDelete;
     }
 
     @Override
-    public int indexOf(Object o) {
-        //TODO
-        return 0;
+    public int indexOf(Object t) {
+        if (Objects.isNull(t)) {
+            for (int i = 0; i < size; i++) {
+                if (null == elements[i]) {
+                    return i;
+                }
+            }
+        }
+        for (int i = 0; i < size; i++) {
+            if (t.equals(elements[i])) {
+                return i;
+            }
+        }
+        return -1;
     }
 
     @Override
-    public int lastIndexOf(Object o) {
-        //TODO
-        return 0;
+    public int lastIndexOf(Object t) {
+        if (Objects.nonNull(t)) { // ???
```

Comment:
what happens if it is null ?

Reply:
Fixed, if it will be "null" it will give us the position of this "null".